src/realm/util/serializer.cpp

```diff
@@ -164,16 +165,24 @@ std::string print_value<>(realm::ObjKey k)
     return ss.str();
 }
 
-template <>
-std::string print_value<>(realm::ObjLink link)
+std::string print_value(realm::ObjLink link, Group* g)
 {
-    std::stringstream ss;
     if (!link) {
-        ss << "NULL";
+        return "NULL";
     }
     else {
-        ss << "L" << link.get_table_key().value << ":" << link.get_obj_key().value;
+        TableRef target_table = g->get_table(link.get_table_key());
+        if (ColKey pk_col = target_table ? target_table->get_primary_key_column() : ColKey{}) {
+            if (auto obj = target_table->try_get_object(link.get_obj_key())) {
```

Comment:
I notice a problem here which is not trivial to solve:
1) User A makes a subscription which queries to a link that exists locally.
2) User B deletes the target object.
3) User A adds another unrelated query to the subscription set.
4) Realm serializes User A's subscription queries again, but not finding the original target object locally here uses the `Lx:y` syntax.
5) Server cannot parse this and sends a subscription error.

Reply:
We could refuse to serialize this query.

Reply:
What do you think about giving ObjLink an `util::Optional<Mixed> m_pk` member field? Then the serializer would use that info to print the primary key directly without looking up the object key which will work even if the object is deleted.

Reply:
That would be a hack, I think. But are you sure the scenario will play out as described above? As far as I can tell, the query string is stored as text in the file. Will the query ever be serialized again?

Reply:
Ok, you are correct. A subscription query is serialized once and the text is stored and reused. So maybe this is not a problem after all. @jbreams can you confirm this?

Reply:
Confirmed.